README.md
# videojs-responsive-controls

Adapts Video.js controls to different player sizes.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
## Table of contents

- [Installation](#installation)
- [Basics](#basics)
- [Defaults](#defaults)
- [Features](#features)
    - [Visible by default](#visible-by-default)
    - [Specifying own breakpoints](#specifying-own-breakpoints)
    - [Using with legacy plugins](#using-with-legacy-plugins)
- [Bundling with legacy systems](#bundling-with-legacy-systems)
  - [`<script>` Tag](#script-tag)
  - [Browserify/CommonJS](#browserifycommonjs)
  - [RequireJS/AMD](#requirejsamd)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
## Installation

```sh
npm install --save videojs-responsive-controls
```


## Basics

Basic usage:

```js
import 'videojs-responsive-controls';

player.responsiveControls({
    controls: {
      remainingTimeDisplay: { mini: false, small: false },
      'vjs-resolution-button': { mini: false },
      'vjs-language-container': { mini: false }
    }
});
```

Allows controlling both Video.js native plugins and legacy plugins (via their class names).

## Defaults

Default breakpoints are:

- `mini`: <450px,
- `small`: <600px,
- `default`: any higher resolution

Default settings are:

```
currentTimeDisplay: {
  mini: false,
},
timeDivider: {
  mini: false,
},
durationDisplay: {
  mini: false,
},
remainingTimeDisplay: {
  mini: false,
  small: false,
},
captionsButton: {
  mini: false,
},
```

## Features

#### Visible by default

If user did not specified visibility for given plugin/breakpoint, and the visibility is also
not specified in default settings - the plugin will be visible.

#### Specifying own breakpoints

You can specify own breakpoints like this:

```js
player.responsiveControls({
    sizes: {
        custom: 1024,
    },
    controls: {
        remainingTimeDisplay: { custom: false }
    }
});
```

Default breakpoints will be overwritten by this setting, therefore if you wish to use defaults too
you'll have to specify default breakpoints along with your custom ones like so:


```js
player.responsiveControls({
    sizes: {
        mini: 450,
        small: 600,
        custom: 1024,
    },
    controls: {
        remainingTimeDisplay: { custom: false }
    }
});
```

Breakpoints are mobile-up so specified value is a maximum size video player can have to be affected by this breakpoint.

To hide element by default use `default` breakpoint:

```js
player.responsiveControls({
    controls: {
        remainingTimeDisplay: { default: false, mini: true }
    }
});
```


#### Using with legacy plugins

Legacy plugins does not register in `videojs.controlBar` space but you can still manage them throught
this script passing class name in options, like this:

```js
player.responsiveControls({
    controls: {
        'vjs-language-container': { mini: false }
    }
});
```



## Bundling with legacy systems

To include videojs-responsive-controls on your website or web application, use any of the following legacy methods.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-responsive-controls.min.js"></script>
<script>
  var player = videojs('my-video');

  player.responsiveControls();
</script>
```

### Browserify/CommonJS

When using with Browserify, install videojs-responsive-controls via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-responsive-controls');

var player = videojs('my-video');

player.responsiveControls();
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-responsive-controls'], function(videojs) {
  var player = videojs('my-video');

  player.responsiveControls();
});
```

## License

MIT. Copyright (c) Tomasz Janiczek &lt;tjaniczek@twig-world.com&gt;


[videojs]: http://videojs.com/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2017-06-22
### Changed
- Safari 9 support.
- Bugfixes


test/plugin.test.js
import document from 'global/document';

import QUnit from 'qunit';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/plugin';

const Player = videojs.getComponent('Player');

const classList = pluginInstance => pluginInstance.el().classList;
const isHidden = pluginInstance => classList(pluginInstance).contains('vjs-hidden');
const isVisible = pluginInstance => !isHidden(pluginInstance);
const dispatchResizeEvent = (player, size, clock) => {
  player.el().style.width = `${size}px`;

  const event = document.createEvent('HTMLEvents');

  event.initEvent('scroll', true, false);

  const detector = player.el().querySelectorAll('.erd_scroll_detection_container')[0];
  const detectNode = detector.childNodes[0].childNodes[0].childNodes[0];

  detectNode.dispatchEvent(event);
  clock.tick(1);
};

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-responsive-controls', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

QUnit.test('registers itself with video.js', function(assert) {
  assert.expect(2);

  assert.strictEqual(
    typeof Player.prototype.responsiveControls,
    'function',
    'videojs-responsive-controls plugin was registered'
  );

  this.player.responsiveControls();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);

  assert.ok(
    this.player.hasClass('vjs-responsive-controls'),
    'the plugin adds a class to the player'
  );
});

QUnit.test('uses default settings if no settings are passed', function(assert) {
  assert.expect(5);

  this.player.width(1000);
  this.player.responsiveControls();
  this.clock.tick(2);

  const {
    currentTimeDisplay,
    timeDivider,
    durationDisplay,
    remainingTimeDisplay,
    captionsButton
  } = this.player.controlBar;

  assert.ok(isVisible(currentTimeDisplay));
  assert.ok(isVisible(timeDivider));
  assert.ok(isVisible(durationDisplay));
  assert.ok(isVisible(remainingTimeDisplay));
  assert.ok(isVisible(captionsButton));

});

QUnit.test('uses default settings for different breakpoints', function(assert) {
  assert.expect(5);

  this.player.width(400);
  this.player.responsiveControls();
  this.clock.tick(2);

  const {
    currentTimeDisplay,
    timeDivider,
    durationDisplay,
    remainingTimeDisplay,
    captionsButton
  } = this.player.controlBar;

  assert.ok(isHidden(currentTimeDisplay));
  assert.ok(isHidden(timeDivider));
  assert.ok(isHidden(durationDisplay));
  assert.ok(isHidden(remainingTimeDisplay));
  assert.ok(isHidden(captionsButton));
});

QUnit.test('shows and hides plugins depending on video player size', function(assert) {
  assert.expect(2);

  this.player.width(1000);
  this.player.responsiveControls();
  this.clock.tick(2);

  const { currentTimeDisplay } = this.player.controlBar;

  assert.ok(isVisible(currentTimeDisplay));

  this.clock.tick(2);
  dispatchResizeEvent(this.player, 300, this.clock);

  assert.ok(isHidden(currentTimeDisplay));
});

QUnit.test('allows redefining media queries', function(assert) {
  assert.expect(6);

  this.player.width(1000);
  this.player.responsiveControls({
    sizes: {
      custom: 100
    },
    controls: {
      captionsButton: {
        custom: false
      }
    }
  });

  this.clock.tick(2);

  const {
    remainingTimeDisplay,
    captionsButton
  } = this.player.controlBar;

  assert.ok(isVisible(remainingTimeDisplay));
  assert.ok(isVisible(captionsButton));

  this.clock.tick(2);
  // Below default media query but above custom
  // Default media query should not apply
  dispatchResizeEvent(this.player, 300, this.clock);

  assert.ok(isVisible(remainingTimeDisplay));
  assert.ok(isVisible(captionsButton));

  this.clock.tick(2);
  dispatchResizeEvent(this.player, 100, this.clock);

  assert.ok(isVisible(remainingTimeDisplay));
  assert.ok(isHidden(captionsButton));
});

QUnit.test('allows redefining default value for a single control element', function(assert) {
  assert.expect(1);

  this.player.width(1000);
  this.player.responsiveControls({
    controls: {
      captionsButton: {
        default: false
      }
    }
  });

  this.clock.tick(2);
  const { captionsButton } = this.player.controlBar;

  assert.ok(isHidden(captionsButton));
});

QUnit.test('allows redefining behaviour for single controls', function(assert) {
  assert.expect(6);

  this.player.width(1000);
  this.player.responsiveControls({
    controls: {
      captionsButton: {
        mini: true,
        small: false,
        default: true
      },
      remainingTimeDisplay: {
        mini: false,
        small: true,
        default: false
      }
    }
  });

  this.clock.tick(2);

  const {
    remainingTimeDisplay,
    captionsButton
  } = this.player.controlBar;

  // Default breakpoint
  assert.ok(isHidden(remainingTimeDisplay));
  assert.ok(isVisible(captionsButton));

  this.clock.tick(2);
  dispatchResizeEvent(this.player, 600, this.clock);

  // Mobile Breakpoint
  assert.ok(isVisible(remainingTimeDisplay));
  assert.ok(isHidden(captionsButton));

  // Mini breakpoint
  this.clock.tick(2);
  dispatchResizeEvent(this.player, 450, this.clock);

  assert.ok(isHidden(remainingTimeDisplay));
  assert.ok(isVisible(captionsButton));

});

QUnit.test('allows redefining behaviour for non standard elements', function(assert) {
  assert.expect(2);

  const custom = document.createElement('div');

  custom.className = 'vjs-language-container';
  this.player.el().querySelectorAll('.vjs-control-bar')[0].appendChild(custom);

  this.player.width(400);
  this.player.responsiveControls({
    controls: {
      'vjs-language-container': {
        mini: false,
        small: false
      }
    }
  });

  this.clock.tick(2);
  const customSelector = this.player.el().querySelectorAll('.vjs-language-container')[0];

  assert.ok(customSelector.classList.contains('vjs-hidden'));

  this.clock.tick(2);
  dispatchResizeEvent(this.player, 900, this.clock);
  assert.ok(!customSelector.classList.contains('vjs-hidden'));
});


scripts/umd.rollup.config.js
/**
 * Rollup configuration for packaging the plugin in a module that is consumable
 * as the `src` of a `script` tag or via AMD or similar client-side loading.
 *
 * This module DOES include its dependencies.
 */
import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import json from 'rollup-plugin-json';
import resolve from 'rollup-plugin-node-resolve';

export default {
  moduleName: 'videojsResponsiveControls',
  entry: 'src/plugin.js',
  dest: 'dist/videojs-responsive-controls.js',
  format: 'umd',
  external: [
    'video.js'
  ],
  globals: {
    'video.js': 'videojs'
  },
  legacy: true,
  plugins: [
    resolve({
      browser: true,
      main: true,
      jsnext: true
    }),
    json(),
    commonjs({
      sourceMap: false
    }),
    babel({
      babelrc: false,
      exclude: 'node_modules/**',
      presets: [
        ['es2015', {
          loose: true,
          modules: false
        }]
      ],
      plugins: [
        // 'external-helpers',
        'transform-object-assign'
      ]
    })
  ]
};


scripts/test.rollup.config.js
/**
 * Rollup configuration for packaging the plugin in a test bundle.
 *
 * This includes all dependencies for both the plugin and its tests.
 */
import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import json from 'rollup-plugin-json';
import multiEntry from 'rollup-plugin-multi-entry';
import resolve from 'rollup-plugin-node-resolve';

export default {
  moduleName: 'videojsResponsiveControlsTests',
  entry: 'test/**/*.test.js',
  dest: 'test/dist/bundle.js',
  format: 'iife',
  external: [
    'qunit',
    'qunitjs',
    'sinon',
    'video.js'
  ],
  globals: {
    'qunit': 'QUnit',
    'qunitjs': 'QUnit',
    'sinon': 'sinon',
    'video.js': 'videojs'
  },
  legacy: true,
  plugins: [
    multiEntry({
      exports: false
    }),
    resolve({
      module: true,
      browser: true,
      main: true,
      jsnext: true
    }),
    json(),
    commonjs({
      sourceMap: false
    }),
    babel({
      babelrc: false,
      exclude: 'node_modules/**',
      presets: [
        ['es2015', {
          loose: true,
          modules: false
        }]
      ],
      plugins: [
        'external-helpers',
        'transform-object-assign'
      ]
    })
  ]
};


scripts/modules.rollup.config.js
/**
 * Rollup configuration for packaging the plugin in a module that is consumable
 * by either CommonJS (e.g. Node or Browserify) or ECMAScript (e.g. Rollup).
 *
 * These modules DO NOT include their dependencies as we expect those to be
 * handled by the module system.
 */
import babel from 'rollup-plugin-babel';
import json from 'rollup-plugin-json';

export default {
  moduleName: 'videojsResponsiveControls',
  entry: 'src/plugin.js',
  external: [
    'video.js',
    'element-resize-detector',
    'object.entries'
  ],
  globals: {
    'video.js': 'videojs'
  },
  legacy: true,
  plugins: [
    json(),
    babel({
      babelrc: false,
      exclude: 'node_modules/**',
      presets: [
        ['es2015', {
          loose: true,
          modules: false
        }]
      ],
      plugins: [
        'external-helpers',
        'transform-object-assign'
      ]
    })
  ],
  targets: [
    {dest: 'dist/videojs-responsive-controls.cjs.js', format: 'cjs'},
    {dest: 'dist/videojs-responsive-controls.es.js', format: 'es'}
  ]
};


src/plugin.js
import videojs from 'video.js';
import elementResizeDetectorMaker from 'element-resize-detector';
import entries from 'object.entries';
import {version as VERSION} from '../package.json';

// Default options for the plugin.
const defaults = {
  sizes: {
    mini: 450,
    small: 600
  },
  controls: {
    currentTimeDisplay: {
      mini: false
    },
    timeDivider: {
      mini: false
    },
    durationDisplay: {
      mini: false
    },
    remainingTimeDisplay: {
      mini: false,
      small: false
    },
    captionsButton: {
      mini: false
    }
  }
};

// // Cross-compatibility for Video.js 5 and 6.
const registerPlugin = videojs.registerPlugin || videojs.plugin;
// // const dom = videojs.dom || videojs;

/**
 * Ascending sorting for breakpoints
 *
 * @param      {Array}  previousSize  Previous breakpoint to sort
 * @param      {Array}  nextSize      Next breakpoint to sort
 * @return     {Array}  Breakpoints array sorted in ascending order
 */
const ascending = (previousSize, nextSize) => previousSize[1] - nextSize[1];

/**
 * Gets all active breakpoints that may be applied to current video player size.
 *
 * @param      {number}    playerSize  The player size.
 * @param      {Object}    sizes       Object containing all possible breakpoints.
 * @return     {Function}  The active breakpoints.
 */
const getActiveBreakpoints = (playerSize, sizes) => {
  if (!Object.entries) {
    entries.shim();
  }

  return Object.entries(sizes)
    .sort(ascending)
    .filter(size => size[1] >= playerSize);
};

/**
 * Gets the breakpoint name for the current player's size
 *
 * @param      {Player}   player  Video.js player object.
 * @param      {Object}   sizes   Object containing all possible breakpoints.
 * @return     {string}   Breakpoint name.
 */
const getPlayerSize = (player, sizes) => {
  const playerSize = player.el().clientWidth;
  const breakpoints = getActiveBreakpoints(playerSize, sizes);

  return breakpoints[0] ? breakpoints[0][0] : 'default';
};

/**
 * Controls should be visible by default so this function checks if
 * desired state was specified and returns either desired state or true.
 *
 * @param      {boolean}  setting  Desired state.
 * @return     {boolean}  Desired state or true.
 */
const trueByDefault = setting => typeof setting !== 'undefined' ? !!setting : true;

/**
 * Determines if given control is a native Video.js plugin by searching for it
 * in player.controlBar.
 *
 * @param      {Player}   player   Video.js player object.
 * @param      {string}   control  Name of the plugin.
 * @return     {boolean}  True if native, False otherwise.
 */
const isNative = (player, control) => typeof player.controlBar[control] === 'object';

/**
 * Uses Video.js API to hide or show Video.js plugin.
 *
 * @param      {Player}   player   Video.js player object.
 * @param      {Object}   control  The element to trigger.
 * @param      {boolean}  show     Desired state.
 * @return     {Function} Control element
 */
const setNative = (player, control, show) => {
  const target = player.controlBar[control];

  return show ? target.show() : target.hide();
};

/**
 * Uses CSS to hide or show custom element. Useful for legacy plugins that
 * do not follow official plugin structure.
 *
 * @param      {Player}     player      Video.js player object.
 * @param      {string}     className   Class of the element to trigger.
 * @param      {boolean}    show        Desired state.
 * @return     {(DOMElement|undefined)} Control element
 */
const setCustom = (player, className, show) => {
  const target = player.el().querySelectorAll(`.${className}`);
  const hiddenClass = 'vjs-hidden';

  if (target.length === 0) {
    return undefined;
  }

  return show ? target[0].classList.remove(hiddenClass) :
                target[0].classList.add(hiddenClass);
};

/**
 * Applies specified settingf to the control element based on video player's size
 *
 * @param      {Player}         player   Video.js player object.
 * @param      {Object|string}  control  Video.js plugin instance or class name.
 * @param      {boolean}        setting  Indites if the control element should be visible
 *                                       on given breakpoint.
 * @return     {(DOMElement|Object)} Control element
 */
const set = (player, control, setting) => {
  const target = trueByDefault(setting);
  const native = isNative(player, control);

  return native ? setNative(player, control, target) :
                  setCustom(player, control, target);
};

/**
 * Applies specified settings to the video player based on it's size
 *
 * @param      {Player}  player         Video.js player object.
 * @param      {Object}  arg2           Settings object
 * @param      {Object}  arg2.sizes     Breakpoints on which the controls will update.
 * @param      {Object}  arg2.controls  Specifies states of all the controls
 *                                      for different breakpoints.
 * @return     {undefined}
 */
const setup = (player, { sizes, controls }) => {
  const size = getPlayerSize(player, sizes);

  for (const control in controls) {
    const setting = controls[control][size];

    set(player, control, setting);
  }
};

/**
 * Initializes new instance of resize detector
 *
 * @return     {ElementResizeDetector} Detector instance
 */
const newDetector = () => {
  return elementResizeDetectorMaker({ strategy: 'scroll' });
};

/**
 * Overwrites default breakpoints with ones specified by the developer.
 *
 * @param      {Object}  settings  Settings passed on initialisation.
 * @param      {Object}  defaultSettings  Default settings for plugin.
 * @return     {Object}  Settings to be applied.
 */
const getMediaQueries = (settings, defaultSettings) => (
  settings && settings.sizes ? settings.sizes : defaultSettings.sizes
);

/**
 * Uses deep-merge to merge default controls settings with ones
 * specified by the developer.
 *
 * @param      {Object}  settings  Settings paased on initialisation.
 * @param      {Object}  defaultSettings  Default settings for plugin.
 * @return     {Object}  Settings to be applied.
 */
const mergeUserSettings = (settings, defaultSettings) => (
  settings && settings.controls ?
  videojs.mergeOptions(defaultSettings.controls, settings.controls) :
  defaultSettings.controls
);

/**
 * A video.js plugin.
 *
 * In the plugin function, the value of `this` is a video.js `Player`
 * instance. You cannot rely on the player being in a "ready" state here,
 * depending on how the plugin is invoked. This may or may not be important
 * to you; if not, remove the wait for "ready"!
 *
 * @function responsiveControls
 * @param    {Object} [userSettings={}]
 *           An object of options left to the plugin author to define.
 */
const responsiveControls = function(userSettings) {
  const settings = {
    sizes: getMediaQueries(userSettings, defaults),
    controls: mergeUserSettings(userSettings, defaults)
  };

  this.addClass('vjs-responsive-controls');
  this.resizeDetector = this.resizeDetector || newDetector();
  this.resizeDetector.listenTo(this.el(), () => setup(this, settings));
};

// Register the plugin with video.js.
registerPlugin('responsiveControls', responsiveControls);

// Include the version number.
responsiveControls.VERSION = VERSION;

export default responsiveControls;
